engine/src/main/java/org/pentaho/di/trans/step/BaseStep.java

```diff
@@ -1310,44 +1313,42 @@ private void handlePutRow( RowMetaInterface rowMeta, Object[] row ) throws Kettl
       }
     }
 
-    if ( outputRowSets.isEmpty() ) {
-      // No more output rowsets!
-      // Still update the nr of lines written.
-      //
-      incrementLinesWritten();
+    synchronized ( outputRowSetsLock ) {
+      if ( outputRowSets.isEmpty() ) {
+        // No more output rowsets!
+        // Still update the nr of lines written.
+        //
+        incrementLinesWritten();
 
-      return; // we're done here!
-    }
+        return; // we're done here!
+      }
 
-    // Repartitioning happens when the current step is not partitioned, but the next one is.
-    // That means we need to look up the partitioning information in the next step..
-    // If there are multiple steps, we need to look at the first (they should be all the same)
-    //
-    switch ( repartitioning ) {
-      case StepPartitioningMeta.PARTITIONING_METHOD_NONE:
-        noPartitioning( rowMeta, row );
-        break;
+      // Repartitioning happens when the current step is not partitioned, but the next one is.
+      // That means we need to look up the partitioning information in the next step..
+      // If there are multiple steps, we need to look at the first (they should be all the same)
+      //
+      switch ( repartitioning ) {
+        case StepPartitioningMeta.PARTITIONING_METHOD_NONE:
+          noPartitioning( rowMeta, row );
+          break;
 
-      case StepPartitioningMeta.PARTITIONING_METHOD_SPECIAL:
-        specialPartitioning( rowMeta, row );
-        break;
-      case StepPartitioningMeta.PARTITIONING_METHOD_MIRROR:
-        mirrorPartitioning( rowMeta, row );
-        break;
-      default:
-        throw new KettleStepException( "Internal error: invalid repartitioning type: " + repartitioning );
+        case StepPartitioningMeta.PARTITIONING_METHOD_SPECIAL:
+          specialPartitioning( rowMeta, row );
+          break;
+        case StepPartitioningMeta.PARTITIONING_METHOD_MIRROR:
+          mirrorPartitioning( rowMeta, row );
+          break;
+        default:
+          throw new KettleStepException( "Internal error: invalid repartitioning type: " + repartitioning );
+      }
     }
   }
 
   /**
    * Copy always to all target steps/copies
-   *
-   * @param rowMeta
-   * @param row
    */
   private void mirrorPartitioning( RowMetaInterface rowMeta, Object[] row ) {
-    for ( int r = 0; r < outputRowSets.size(); r++ ) {
-      RowSet rowSet = outputRowSets.get( r );
+    for ( RowSet rowSet : outputRowSets ) {
```

Comment:
Seems to me, the outputRowSets can mutate during iteration. So I'd think a this would need a readLock at this point.

Reply:
This is private extracted method only used in synchronized context. I see the only reason to add synchronization here - to make the code less fragile. Should I?